triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -737,23 +746,62 @@ class Runner(
         // any other error will cause the trigger's stream to fail
       }
     }
-    import io.grpc.Status.Code
-    import Code.RESOURCE_EXHAUSTED
-    def retryableSubmit(req: SubmitRequest) =
+
+    def retryableSubmit(req: SubmitRequest): Future[Option[Option[SingleCommandFailure]]] =
       submit(req).map {
         case Some(SingleCommandFailure(_, s))
             if (s.getStatus.getCode: Code) == (RESOURCE_EXHAUSTED: Code) =>
           None
         case z => Some(z)
       }
-    retrying(
-      initialTries = maxTriesWhenOverloaded,
-      backoff = overloadedRetryDelay,
-      parallelism = maxParallelSubmissionsPerTrigger,
-      retryableSubmit,
-      submit,
-    )
-      .collect { case Some(err) => err }
+
+    val throttleFlow: Flow[SubmitRequest, SubmitRequest, NotUsed] =
+      Flow[SubmitRequest]
+        .throttle(maxSubmitRequests, maxSubmitRequestsDuration)
+    val submitFlow: Flow[SubmitRequest, SingleCommandFailure, NotUsed] =
+      Flow[SubmitRequest]
+        .filter { _ =>
+          pendingCommandIds.count(_._2 == SeenMsgs.Neither) <= maxInFlightCommands
+        }
+        .via(
+          retrying(
+            initialTries = maxTriesWhenOverloaded,
+            backoff = overloadedRetryDelay,
+            parallelism = maxParallelSubmissionsPerTrigger,
+            retryableSubmit,
+            submit,
+          )
+            .collect { case Some(err) => err }
+        )
```

Comment:
This flow (edge) performs command submissions only when pending commands are below `maxInFlightCommands`